File: 41. First Missing Positive/41.cpp
class Solution
{
public:
    int firstMissingPositive(vector<int> &nums)
    {
        int n = nums.size();
        if (n == 0)
            return 1;
        for (int i = 0; i < n; i++)
        {
            while (nums[i] > 0 && nums[i] <= n && nums[i] != nums[nums[i] - 1]) //大于0，小于等于n，不等于自己
            {
                //交换位置
                int tmp = nums[i];
                nums[i] = nums[tmp - 1];
                nums[tmp - 1] = tmp;
            }
        }
        for (int i = 0; i < n; i++)
        {
            if (nums[i] != i + 1) //找到第一个不等于自己的数
                return i + 1;
        }
        return n + 1; //所有数都等于自己，即1,2,3,...,n，返回n+1
    }
};

File: 41. First Missing Positive/41.kt
class Solution {
    fun firstMissingPositive(nums: IntArray): Int {
        val n = nums.size
        if (n == 0) return 1
        for (i in 0 until n) {
            while (nums[i] > 0 && nums[i] <= n && nums[i] != nums[nums[i] - 1]) {
                val temp = nums[i]
                nums[i] = nums[temp - 1]
                nums[temp - 1] = temp
            }
        }
        for (i in 0 until n) {
            if (nums[i] != i + 1) return i + 1
        }
        return n + 1
    }
}
